crates/oxc_query/src/vertex.rs

```diff
@@ -0,0 +1,320 @@
+            Self::VariableDeclaration(data) => data.variable_declaration.span,
+            Self::ReturnStatementAST(data) => data.return_statement.span,
+            Self::File
+            | Self::URL(_)
+            | Self::PathPart(_)
+            | Self::SearchParameter(_)
+            | Self::Span(_) => {
+                unreachable!("Tried to get the span from a {self:#?}")
+            }
+        }
+    }
+
+    pub fn ast_node_id(&self) -> Option<AstNodeId> {
+        match &self {
+            Vertex::ASTNode(data) => Some(data.id()),
+            Vertex::Class(data) => data.ast_node.map(|x| x.id()),
+            Vertex::Import(data) => data.ast_node.map(|x| x.id()),
+            Vertex::Interface(data) => data.ast_node.map(|x| x.id()),
+            Vertex::JSXElement(data) => data.ast_node.map(|x| x.id()),
+            Vertex::TypeAnnotation(data) => data.ast_node.map(|x| x.id()),
+            Vertex::VariableDeclaration(data) => data.ast_node.map(|x| x.id()),
+            Vertex::ReturnStatementAST(data) => data.ast_node.map(|x| x.id()),
+            Vertex::DefaultImport(_)
+            | Vertex::AssignmentType(_)
+            | Vertex::ClassMethod(_)
+            | Vertex::Expression(_)
+            | Vertex::File
+            | Vertex::InterfaceExtend(_)
+            | Vertex::JSXAttribute(_)
+            | Vertex::JSXExpressionContainer(_)
+            | Vertex::JSXFragment(_)
+            | Vertex::ObjectLiteral(_)
+            | Vertex::JSXText(_)
+            | Vertex::JSXSpreadChild(_)
+            | Vertex::JSXSpreadAttribute(_)
+            | Vertex::JSXOpeningElement(_)
+            | Vertex::PathPart(_)
+            | Vertex::URL(_)
+            | Vertex::Type(_)
+            | Vertex::SpecificImport(_)
+            | Vertex::Span(_)
+            | Vertex::SearchParameter(_)
+            | Vertex::ClassProperty(_) => None,
+        }
+    }
+
+    pub fn make_url(attr: &'a JSXAttribute<'a>) -> Option<Self> {
+        let Some(maybe_url) = jsx_attribute_to_constant_string(attr) else { return None };
+        let Ok(parsed_url) = Url::parse(&maybe_url) else { return None };
+        return Some(Vertex::URL(Rc::new(parsed_url)));
+    }
+
+    pub fn as_constant_string(&self) -> Option<String> {
+        match &self {
+            Vertex::Expression(expr) => expr_to_maybe_const_string(expr),
+            _ => None,
+        }
+    }
+}
+
+impl Typename for Vertex<'_> {
+    fn typename(&self) -> &'static str {
+        match self {
+            Vertex::ASTNode(_) => "ASTNode",
+            Vertex::AssignmentType(_) => "AssignmentType",
+            Vertex::Class(class) => class.typename(),
+            Vertex::ClassMethod(_) => "ClassMethod",
+            Vertex::ClassProperty(_) => "ClassProperty",
+            Vertex::DefaultImport(_) => "DefaultImport",
+            Vertex::Expression(_) => "Expression",
+            Vertex::File => "File",
+            Vertex::Import(import) => import.typename(),
```

Comment:
I wish there was some way to say Vertex::*(x: impl Typename) => x.typename() or some way to enforce this since this will definitely cause some bugs at some point I think.

Reply:
this was just the cause of another bug, this would be incredibly helpful to implement somehow

Reply:
Unfortunately all the ways I can think of involve this kind of `match self` expression _somewhere_ =/